### PR TITLE
merge: (#16) Aws S3 설계하기

### LINKED ISSUE #16: ThirdParty 구축하기
### Describe

* S3
  - FileOperationUtil
  - FilePersistenceAdapter
  - FilePort

* File
  - FileEntity
  - FileRepository

### Additional

_No response_

### PR DESCRIPTION
**Pull request by khcho0125**
close #16

### CODE REVIEW

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/feed/FeedPersistenceAdapter.kt [hunk 1]
```diff
@@ -3,20 +3,30 @@ package comit.simsimpulyi.koala.persistence.feed
...
+    private val feedRepository: FeedRepository,
+    private val filePersistenceAdapter: FilePersistenceAdapter,
+    private val feedMapper: FeedMapper
 ) : SaveFeedPort, QueryFeedPort {
 
     override fun saveFeed(feed: Feed): Feed {
-        TODO("Not yet implemented")
+        var entity = feedMapper.toEntity(feed)
+
+        entity = feedRepository.save(entity)
```
Comment (softpeanut) [inline koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/feed/FeedPersistenceAdapter.kt]: 로직 하나로 묶어서 val로 하는게 좋을 것 같아요

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/feed/FeedPersistenceAdapter.kt [hunk 2]
```diff
@@ -3,20 +3,30 @@ package comit.simsimpulyi.koala.persistence.feed
...
+        filePersistenceAdapter.saveFile(feed.filePaths, entity)
+
+        return feedMapper.toDomain(entity)
     }
 
-    override fun getFeed(feedId: UUID): Feed {
-        TODO("Not yet implemented")
+    override fun queryFeed(feedId: UUID): Feed {
+        val entity = feedRepository.queryById(feedId).orElseThrow() // exception 추가
```
Comment (softpeanut) [inline koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/feed/FeedPersistenceAdapter.kt]: 코틀린에서는 Optional 안씁니다

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/FilePersistenceAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package comit.simsimpulyi.koala.persistence.file
+
+import comit.simsimpulyi.koala.persistence.feed.entity.FeedEntity
+import comit.simsimpulyi.koala.persistence.file.entity.FileEntity
+import org.springframework.stereotype.Component
+import java.util.*
+
+@Component
+class FilePersistenceAdapter(
```
Comment (softpeanut) [inline koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/FilePersistenceAdapter.kt:9]: 포트 상속 안받나요?

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/FilePersistenceAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    private val fileRepository: FileRepository
+) {
+
+    fun saveFile(filePaths: List<String>, feed: FeedEntity) {
+        for (path in filePaths) {
+            fileRepository.save(
```
Comment (softpeanut) [inline koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/FilePersistenceAdapter.kt]: for 문으로 save 하는 것보다 saveAll하는게 좋을 것 같아요

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/entity/FileEntity.kt [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+import comit.simsimpulyi.koala.persistence.BaseTimeEntity
+import comit.simsimpulyi.koala.persistence.feed.entity.FeedEntity
+import javax.persistence.*
+import javax.validation.constraints.NotNull
+
+@Entity
+@Table(name = "tbl_file")
+class FileEntity(
+    @Id
```
Comment (softpeanut) [inline koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/entity/FileEntity.kt]: FileEntity의 id로 path를 사용하면 안될 것 같습니다
혹시 모를 상황을 대비해 별도 식별자 필드를 두고 path는 일반 컬럼으로 만들어야 할 것 같습니다

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/feed/mapper/FeedMapper.kt [hunk 6]
```diff
@@ -3,20 +3,25 @@ package comit.simsimpulyi.koala.persistence.feed.mapper
...
+
+    @Autowired
+    protected lateinit var filePersistenceAdapter: FilePersistenceAdapter
 
     @Mapping(target = "writerId", expression = "java(entity.getWriter().getId())")
+    @Mapping(target = "filePaths", expression = "java(filePersistenceAdapter.queryFileList(entity.getId()))")
     abstract override fun toDomain(entity: FeedEntity): Feed
 
-    @Mapping(target = "writer", expression = "java(userPersistenceAdapter.getUser(model.getWriterId()))")
+    @Mapping(target = "writer", expression = "java(userRepository.queryUserEntityById(model.getWriterId()))")
```
Comment (softpeanut) [inline koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/feed/mapper/FeedMapper.kt]: 저번에도 말했듯이 userRepository를 직접사용하는 것이 아닌 UserPersistenceAdapter를 통해 변환해야 할 것 같습니다

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/FilePersistenceAdapter.kt (hunk 3 above)
Comment (khcho0125) [reply koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/FilePersistenceAdapter.kt:9]: 같은 계층이라도 필요한 건가요?

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/feed/mapper/FeedMapper.kt (hunk 6 above)
Comment (khcho0125) [reply koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/feed/mapper/FeedMapper.kt]: 어댑터를 거치면 도메인 모델로 되어 Repository를 사용했는데
어댑터에도 엔티티를 반환하는 쿼리를 추가하겠습니다

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/FilePersistenceAdapter.kt (hunk 3 above)
Comment (softpeanut) [reply koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/FilePersistenceAdapter.kt:9]: FeedPersistenceAdapter에서만 사용되는 거라면 필요 없을 것 같습니다

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/FilePersistenceAdapter.kt (hunk 4 above)
Comment (khcho0125) [reply koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/FilePersistenceAdapter.kt]: <html>
<body>
<!DOCTYPE html><p cid="n0" mdtype="paragraph" style="box-sizing: border-box; -webkit-margin-before: 1rem; -webkit-margin-after: 1rem; -webkit-margin-start: 0px; -webkit-margin-end: 0px; margin: 30px 0px 0.8em; width: inherit; position: relative; color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: pre-wrap; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-style: initial; text-decoration-color: initial;"><span class="md-line md-end-block md-focus" cid="n2" mdtype="line" style="box-sizing: border-box; display: block;"><span md-inline="plain" class="md-expand" style="box-sizing: border-box;">saveAll을 사용할 때 map을 사용해 변환 후 save 됩니다.</span></span></p><pre class="md-fences md-end-block" lang="kotlin" contenteditable="false" cid="n8" mdtype="fences" style="box-sizing: border-box; overflow: visible; font-family: monospace, monospace; font-size: 0.9em; white-space: pre; display: block; break-inside: avoid; text-align: left; background-image: ; background-position: var(--code-block-bg-color); background-size: ; background-repeat: var(--code-block-bg-color); background-attachment: ; background-origin: ; background-clip: ; background-color: rgb(248, 248, 248); position: relative !important; border: 1px solid rgb(231, 234, 237); border-radius: 3px; padding: 8px 4px 6px; margin-bottom: 15px; margin-top: 15px; width: inherit; color: rgb(51, 51, 51); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: 2; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-style: initial; text-decoration-color: initial;"><span style="box-sizing: border-box;"></span><br><span role="presentation" style="box-sizing: border-box; padding-right: 0.1px;"><span class="cm-variable" style="box-sizing: border-box; color: black;">filePaths</span>.<span class="cm-variable" style="box-sizing: border-box; color: black;">map</span> { <span class="cm-variable" style="box-sizing: border-box; color: black;">FileEntity</span>(<span class="cm-variable" style="box-sizing: border-box; color: black;">it</span>, <span class="cm-variable" style="box-sizing: border-box; color: black;">feed</span>) }</span></pre><p cid="n132" mdtype="paragraph" style="box-sizing: border-box; -webkit-margin-before: 1rem; -webkit-margin-after: 1rem; -webkit-margin-start: 0px; -webkit-margin-end: 0px; margin: 0.8em 0px; width: inherit; position: relative; color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: pre-wrap; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-style: initial; text-decoration-color: initial;"><span class="md-line md-end-block" cid="n127" mdtype="line" style="box-sizing: border-box; display: block;"><span md-inline="plain" class="" style="box-sizing: border-box;">아래는 십만개의 데이터로 측정한 결과입니다. (nanosecond)</span></span></p><figure class="md-table-fig" contenteditable="false" cid="n89" mdtype="table" style="box-sizing: border-box; margin: -8px 0px 0px -8px; overflow-x: auto; max-width: calc(100% + 16px); padding: 8px; color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: pre-wrap; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-style: initial; text-decoration-color: initial;">

  | 1차 | 2차 | 3차 | 4차 | 5차
-- | -- | -- | -- | -- | --
map | 13578000 | 15744700 | 13493500 | 17532100 | 15994500
for | 10088200 | 8884700 | 13546800 | 7907100 | 8705600
forEach | 6996300 | 9180800 | 9617000 | 13011100 | 7161600

</figure><ul class="ul-list" cid="n123" mdtype="list" data-mark="*" style="box-sizing: border-box; margin: 0.8em 0px; padding-left: 30px; color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: pre-wrap; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-style: initial; text-decoration-color: initial;"><li cid="n124" mdtype="list_item" class="" style="box-sizing: border-box; margin: 0px; position: relative;"><p cid="n125" mdtype="paragraph" style="box-sizing: border-box; -webkit-margin-before: 1rem; -webkit-margin-after: 1rem; -webkit-margin-start: 0px; -webkit-margin-end: 0px; margin: 0.5rem 0px; width: inherit; position: relative;"><span class="md-line md-end-block" cid="n126" mdtype="line" style="box-sizing: border-box; display: block;"><span md-inline="plain" class="" style="box-sizing: border-box;">map 평균 : 15,268,560 ns</span></span></p></li><li cid="n135" mdtype="list_item" class="" style="box-sizing: border-box; margin: 0px; position: relative;"><p cid="n136" mdtype="paragraph" style="box-sizing: border-box; -webkit-margin-before: 1rem; -webkit-margin-after: 1rem; -webkit-margin-start: 0px; -webkit-margin-end: 0px; margin: 0.5rem 0px; width: inherit; position: relative;"><span class="md-line md-end-block" cid="n137" mdtype="line" style="box-sizing: border-box; display: block;"><span md-inline="plain" class="" style="box-sizing: border-box;">for 평균 : 9,826,480 ns</span></span></p></li><li cid="n144" mdtype="list_item" class="" style="box-sizing: border-box; margin: 0px; position: relative;"><p cid="n143" mdtype="paragraph" style="box-sizing: border-box; -webkit-margin-before: 1rem; -webkit-margin-after: 1rem; -webkit-margin-start: 0px; -webkit-margin-end: 0px; margin: 0.5rem 0px; width: inherit; position: relative;"><span class="md-line md-end-block" cid="n138" mdtype="line" style="box-sizing: border-box; display: block;"><span md-inline="plain" class="" style="box-sizing: border-box;">forEach 평균 : 9,193,360 ns</span></span></p></li></ul><p cid="n193" mdtype="paragraph" style="box-sizing: border-box; -webkit-margin-before: 1rem; -webkit-margin-after: 1rem; -webkit-margin-start: 0px; -webkit-margin-end: 0px; margin: 0.8em 0px; width: inherit; position: relative; color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: pre-wrap; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-style: initial; text-decoration-color: initial;"><span class="md-line md-end-block" cid="n194" mdtype="line" style="box-sizing: border-box; display: block;"></span></p><p cid="n209" mdtype="paragraph" style="box-sizing: border-box; -webkit-margin-before: 1rem; -webkit-margin-after: 1rem; -webkit-margin-start: 0px; -webkit-margin-end: 0px; margin: 0.8em 0px; width: inherit; position: relative; color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: pre-wrap; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-style: initial; text-decoration-color: initial;"><span class="md-line md-end-block" cid="n204" mdtype="line" style="box-sizing: border-box; display: block;"><span md-inline="plain" class="" style="box-sizing: border-box;">위의 측정 결과는 List 기준이며 표준 편차가 발생할 수 있습니다.</span></span></p>
</body>
</html>

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/entity/FileEntity.kt (hunk 5 above)
Comment (khcho0125) [reply koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/entity/FileEntity.kt]: path에는 UUID가 사용돼서 중복될 일이 없을 것이라고 생각하는데 기본키로 두면 안되는 이유가 무엇인가요??
확장의 이유인가요 아니면 성능 때문인가요..

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/FilePersistenceAdapter.kt (hunk 4 above)
Comment (softpeanut) [reply koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/FilePersistenceAdapter.kt]: 어플리케이션의 데이터를 단순히 처리하는 성능만 측정하신 건가요?
save 메서드가 호출될 때마다 불필요한 프록시 과정을 거치게 되어 성능이 저하됩니다
saveall 메서드도 프록시 과정을 거치지만 내부에서 this 키워드로 처리하기 때문에 매번 프록시 과정을 거치지 않습니다

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/entity/FileEntity.kt (hunk 5 above)
Comment (softpeanut) [reply koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/entity/FileEntity.kt]: 만약 어딘가에서 사진을 긁어온 사람이 게시글을 올릴 때 똑같은 사진을 2번 사용하게 된다면 중복이 발생하지 않을까요?

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/feed/FeedRepository.kt [hunk 7]
```diff
@@ -7,4 +7,6 @@ import java.util.*
 
 @Repository
 interface FeedRepository : CrudRepository<FeedEntity, UUID> {
+
+    fun queryById(id: UUID): FeedEntity
```
Comment (softpeanut) [inline koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/feed/FeedRepository.kt]: Optional을 사용하지 않는 것이 Null-safe 로직을 작성하지 말라는 소리가 아닙니다...
kotlin의 ? 키워드를 이용해 예외 처리를 하는 것을 말한 겁니다

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/FilePersistenceAdapter.kt (hunk 4 above)
Comment (khcho0125) [reply koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/FilePersistenceAdapter.kt]: save 할 때마다 새로운 트랜잭션이 생성되어 성능이 오히려 안 좋아지는군요..
saveAll로 수정하겠습니다

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/entity/FileEntity.kt (hunk 5 above)
Comment (khcho0125) [reply koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/entity/FileEntity.kt]: 똑같은 사진을 사용해도 path의 uuid가 다르기 때문에 각각 따로 저장이 되는 방식으로 구현을 했습니다.
이름이 같은 파일이 저장될 때 생기는 문제를 배제할 수 없기 때문에 path는 결코 중복될 수 없어야 합니다.

AwsS3Adapter의 fileKey를 이용해 저장할 계획입니다.
```kotlin
    override fun uploadFile(file: File, folder: BucketFolder): String {
        val fileKey = "${folder.name}/${UUID.randomUUID()}/${file.name}"
```

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/feed/FeedRepository.kt (hunk 7 above)
Comment (khcho0125) [reply koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/feed/FeedRepository.kt]: 아 Null을 잊었군요
수정하겠습니다

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/entity/FileEntity.kt (hunk 5 above)
Comment (softpeanut) [reply koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/entity/FileEntity.kt]: 사용자가 반드시 정상적인 경로로 사진을 보낸다는 보장이 없으니 s3에 올라간 url을 2번 보낼 수 있지 않을까요

Comment (khcho0125) [reply koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/entity/FileEntity.kt]: 제가 이해한 대로라면

> 파일을 S3 저장(url 획득) → Feed 저장 요청 (파일 url과 함께) → 파일 url과 함께 Feed 저장

이 경우 중복된 두 경로를 구별하는 방안이 필요하다

제 생각은

> Feed 저장 요청 (파일과 함께) → 파일을 S3 저장 → 파일 url과 함께 Feed 저장

제가 이해한 상황에 따른 문제가 있을까요...?

Comment (softpeanut) [reply koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/persistence/file/entity/FileEntity.kt]: 첫번째로 말한 "방안 파일을 S3 저장(url 획득) → Feed 저장 요청 (파일 url과 함께) → 파일 url과 함께 Feed 저장"이 맞는 것 같습니다

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/thirdparty/aws/s3/AwsS3Adapter.kt [hunk 8]
```diff
@@ -15,8 +14,8 @@ class AwsS3Adapter(
     private val awsS3Properties: AwsS3Properties
 ) : UploadFilePort {
 
-    override fun uploadFile(file: File, folder: BucketFolder): String {
-        val fileKey = "${folder.name}/${UUID.randomUUID()}/${file.name}"
+    override fun uploadFile(file: File): String {
+        val fileKey = "${UUID.randomUUID()}${file.name}"
```
Comment (softpeanut) [inline koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/thirdparty/aws/s3/AwsS3Adapter.kt]: 뭔가 uuid와 file의 이름과 구분이 되었으면 좋겠어요

**Review by softpeanut — DISMISSED** [review]

File: koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/thirdparty/aws/s3/AwsS3Adapter.kt (hunk 8 above)
Comment (khcho0125) [reply koala-infrastructure/src/main/kotlin/comit/simsimpulyi/koala/thirdparty/aws/s3/AwsS3Adapter.kt]: `@` 로 구분하겠습니다

**Review by softpeanut — APPROVED** [review]

**Pull request merged**